merkledag/README.md

```diff
@@ -0,0 +1,144 @@
+- `multihash` - the [multihash](https://github.com/jbenet/multihash) format / protocol.
+- `ipfs object` - a node in the ipfs merkledag. It represents a singular entity.
+- `merkledag format` - the format that ipfs objects are expressed with.
+- `link segment` or `link table` - the part of the merkledag format that expresses links to other objects.
+- `data segment` - the part of the merkledag format that expresses object data.
+- `protobuf` - protocol buffers, a serialization encoding.
+- `multicodec` - a self-describing, generalized serialization format.
+
+## Abstract
+
+The _ipfs merkledag_ is a directed acyclic graph whose edges are merkle-links. This means that links to objects can authenticate the objects themselves, and that every object contains a secure representation of its children.
+
+This is a powerful primitive for distributed systems computations. The merkledag simplifies distributed protocols by providing an append-only authenticated datastructure. Parties can communicate and exchange secure references (merkle-links) to objects. The references are enough to verify the correctness of the object at a later time, which allows the objects themselves to be served over untrusted channels. Merkledags also allow the branching of a datastructure and subsequent merging, as in the version control system git. More generally, merkledags simplify the construction of Secure CRDTs, which enable distributed, convergent, commutative computation in an authenticated, secure way.
+
+## The Format
+
+The IPFS merkledag format is very simple. It serves as a thin waist for more complex applications and data structure transports. Therefore, it aims to be as simple and small as possible.
+
+The format has two parts, the logical format, and the serialized format.
+
+### Logical Format
+
+The merkledag format defines two parts, `Nodes` and `Links` between nodes. `Nodes` embed `Links` in their `Link Segement` (or link table).
+
+A node is divided in two parts:
+- a `Link Segment` which contains all the links.
+- a `Data Segment` which contains the object data.
+
+Instead of following previous approaches to merkledags, which place data mostly at the edges, the IPFS merkledag adapts the format of the HTTP web: every path endpoint is an object with _both_ links and data. (this is fundamentally different from UNIX files, in which objects have _either_ links (directories) _or_ data (files).).
+
+The logical format -- in protobuf -- looks like this:
+
+```proto3
+// An IPFS MerkleDAG Link
+message MerkleLink {
+  bytes Hash = 1;   // multihash of the target object
+  string Name = 2;  // utf string name
+  uint64 Tsize = 3; // cumulative size of target object
+
+  // user extensions start at 50
+}
+
+// An IPFS MerkleDAG Node
+message MerkleNode {
+  repeated MerkleLink Links = 2; // refs to other objects
+  bytes Data = 1; // opaque user data
+
+  // user extensions start at 50
+}
+```
+
+### Serialized Format
+
+
+(TODO remove this? use only protobuf?)
```

Comment:
I think we are okay using only protobuf. since we version tag the repos and the clients and protocols, we can write in logic to handle multicodec OR protobuf later if for some reason we have to switch away from protobuf due to it being proven to be stealing money from the poor, or some other terrible thing.